keras_cv/layers/aspp.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any
+from typing import List
+from typing import Mapping
+
+import tensorflow as tf
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class ASPP(tf.keras.layers.Layer):
+    """Implements the Atrous Spatial Pyramid Pooling.
+
+    References:
+        [Rethinking Atrous Convolution for Semantic Image Segmentation](
+          https://arxiv.org/pdf/1706.05587.pdf)
+    [Encoder-Decoder with Atrous Separable Convolution for Semantic Image
+        Segmentation](https://arxiv.org/pdf/1802.02611.pdf)
+
+    inp = tf.keras.layers.Input((384, 384, 3))
+    backbone = tf.keras.applications.EfficientNetB0(input_tensor=inp, include_top=False)
+    layer_names = ['block2b_add', 'block3b_add', 'block5c_add', 'top_activation']
+
+    backbone_outputs = {}
+    for i, layer_name in enumerate(layer_names):
+        backbone_outputs[i+2] = backbone.get_layer(layer_name).output
+
+    # output_dict is a dict with 4 as keys, since it only process the level 4 backbone
+    # inputs
+    output_dict = keras_cv.layers.ASPP(level=4, dilation_rates=[6, 12, 18])(backbone_outputs)
+
+    # output[4].shape = [None, 16, 16, 256]
+    """
+
+    def __init__(
+        self,
+        level: int,
+        dilation_rates: List[int],
+        num_channels: int = 256,
+        activation: str = "relu",
+        dropout: float = 0.0,
+        **kwargs,
+    ):
+        """Initializes an Atrous Spatial Pyramid Pooling (ASPP) layer.
+
+        Args:
+            level: An `int` level to apply ASPP. This will be used to get the exact input
```

Comment:
I'm a little confused here, would this mean that ASPP can only apply to a single level, while FPN applies to multiple levels? If so, can we still can them both as decoder or spatial pooling? 
It sounds to me that ASPP is more inlined with Inception module, where multiple conv / strides are applied on a single level output., instead of one stride applied on multiple level outputs

Reply:
If you check https://arxiv.org/pdf/1706.05587.pdf (Figure 3), the ASPP is similar as FPN in a sense that its body include the logic for going deep with different dilation rates, whereas the FPN goes deep by using the different level of outputs from the backbone.

Reply:
Comparing Fig 2.b and 2.d, I came to the conclusion that we should probably not call FPN and ASPP to be "similar" and decoder pattern. More specifically, FPN is decoder pattern (2.b), while ASPP is spatial pyramid pooling pattern.
Not related to this PR, but as we have DeepLab models, we might want to differentiate between the two.